src/datasets/arrow_dataset.py

```diff
@@ -1246,43 +1258,118 @@ def save_to_disk(self, dataset_path: str, fs=None):
+                        logger.debug(f"Finished writing shard number {job_id} of {num_shards}.")
+                        shard_lengths[job_id], shard_sizes[job_id] = content
+                    else:
+                        pbar.update(content)
+        with fs.open(path_join(dataset_path, config.DATASET_STATE_JSON_FILENAME), "w", encoding="utf-8") as state_file:
             json.dump(state, state_file, indent=2, sort_keys=True)
         with fs.open(
-            Path(dataset_path, config.DATASET_INFO_FILENAME).as_posix(), "w", encoding="utf-8"
+            path_join(dataset_path, config.DATASET_INFO_FILENAME), "w", encoding="utf-8"
         ) as dataset_info_file:
             # Sort only the first level of keys, or we might shuffle fields of nested features if we use sort_keys=True
             sorted_keys_dataset_info = {key: dataset_info[key] for key in sorted(dataset_info)}
             json.dump(sorted_keys_dataset_info, dataset_info_file, indent=2)
-        logger.info(f"Dataset saved in {dataset_path}")
+
+    @staticmethod
+    def _save_to_disk_single(arg):
+        job_id: Dataset = arg["job_id"]
+        shard: Dataset = arg["shard"]
+        fpath: str = arg["fpath"]
+        storage_options: Optional[dict] = arg["storage_options"]
+        refresh_rate = 0.05  # 20 progress updates per sec
+        batch_size = config.DEFAULT_MAX_BATCH_SIZE
+
+        if shard._indices is not None:
+            raise ValueError(
+                "`_save_to_disk_single` only support shards with flattened indices. "
+                "Please call ds.flatten_indices() before saving to disk."
+            )
+
+        num_examples_progress_update = 0
+        writer = ArrowWriter(
+            features=shard.features,
+            path=fpath,
+            storage_options=storage_options,
+            embed_local_files=True,
+        )
+        try:
+            _time = time.time()
+            if config.PYARROW_VERSION.major >= 8:
+                for pa_table in table_iter(shard.data.table, batch_size=batch_size):
+                    writer.write_table(pa_table)
+                    num_examples_progress_update += len(pa_table)
+                    if time.time() > _time + refresh_rate:
+                        _time = time.time()
+                        yield job_id, False, num_examples_progress_update
+                        num_examples_progress_update = 0
+            else:
+                for i in range(0, shard.num_rows, batch_size):
+                    pa_table = shard.data.slice(i, batch_size)
+                    writer.write_table(pa_table)
+                    num_examples_progress_update += len(pa_table)
+                    if time.time() > _time + refresh_rate:
+                        _time = time.time()
+                        yield job_id, False, num_examples_progress_update
+                        num_examples_progress_update = 0
```

Comment:
I iterate on batches here to update the tqdm bar, but for old versions of pyarrow this may be too slow since `table_iter` only works for pyarrow>=8.

I think we may have to implement `table_iter` even on old versions for performance reasons. It can be based on `pa.Table.to_record_batches` - lmk what you think

Reply:
Ok I just implemented `pa.Table.to_reader` for pyarrow < 8 for our `datasets.table.Table`. This way we don't have to check the pyarrow version anymore